### PR TITLE
WIP for has many releationships with django-filter.

### PR DESCRIPTION
**Pull request by benkonrath**
This shouldn't be merged.

### CODE REVIEW

File: addon/serializers/drf.js [hunk 1]
```diff
@@ -12,6 +12,40 @@ import Ember from 'ember';
...
+      payloadRelKey = this.keyForRelationship(key, relationship.kind);
+      if (!hash.hasOwnProperty(payloadRelKey)) {
+        return;
+      }
+      var adapter = type.store.adapterFor(type);
+
+      if (relationship.kind === 'hasMany' && adapter.useHasManyRelatedFilterUrl) {
+        // FIXME This is the wrong approach. Inverse maybe?
+        var parentTypeKey = relationship.parentType.typeKey;
+        if (parentTypeKey === 'trainerProductGroup') {
```
Comment (benkonrath) [inline addon/serializers/drf.js]: This is hard-coded for one of my projects but it needs to be generalized for it to be useful.

File: addon/serializers/drf.js [hunk 2]
```diff
@@ -12,6 +12,40 @@ import Ember from 'ember';
  */
 export default DS.RESTSerializer.extend({
 
+  normalize: function(type, hash, prop) {
+    this.addRelationshipsToLinks(type, hash);
```
Comment (benkonrath) [inline addon/serializers/drf.js]: To support hyperlink related fields, we could search for URLs on relationships using a regex and add them to the 'links' hash.

**Comment by benkonrath** [comment]
I'd like to implement the hyper link fields feature. It's independent of the the django-filter solution so I can always work on that later. What do you think?

**Comment by dustinfarris** [comment]
Can you share a little of how your serializers are set up?  I'm having some difficulty wrapping my head around this.

**Comment by benkonrath** [comment]
Let me give you a simplified example.

**Product Group**
```python
class ProductGroup(models.Model):
    name = models.CharField(max_length=100)
    products = models.ManyToManyField('Product', related_name='product_groups')

class ProductGroupSerializer(serializers.ModelSerializer):
    class Meta:
        model = ProductGroup
        fields = ('id', 'name', 'products')
```
```
GET: /api/product_groups/23
```
```js
{id: 23, name: 'Fun Products', products: [12, 44, 234]}
```

**Product**
```python
class Product(models.Model):
    name = models.CharField(max_length=100)

class ProductSerializer(serializers.ModelSerializer):
    class Meta:
        model = ProductGroup
        fields = ('id', 'name', 'product_group')
```
```
GET: /api/products/12
```
```js
{id: 12, name: 'Silly string', product_group: 23}
```

I think this is a pretty standard setup so far. The interesting bit is the ProductView with the DjangoFilterBackend.

```python
class ProductViewSet(viewsets.ModelViewSet):
    serializer_class = ProductSerializer
    queryset = Product.objects.all()
    filter_backends = (filters.DjangoFilterBackend,)
    filter_fields = ('product_group',)
```

Now we can make a request for the Products in ProductGroup 23 with this URL:

```
GET: /api/products/?product_group=23
```
```js
[
  {id: 12, name: 'Silly string', product_group: 23},
  {id: 44, name: 'Whoopee cushion', product_group: 23},
  {id: 234, name: 'Stink bomb', product_group: 23},
]
```

The code in this PR just tries to automatically build the  `/api/products/?product_group=23` url. But I think this is too complex to always get right (as you can see by my comment and the fact that I hard-coded some stuff).

It would be better to get DRF to generate the correct URL with a custom HyperLinkedField and just add general support for the HyperLinkedField to EDA. For example, the ProductGroup json would look like this if I wanted to use the django-filter pattern:

```
GET: /api/product_groups/23
```
```js
{id: 23, name: 'Fun Products', products: '/api/products/?product_group=23'}
```

So, the next task on my TODO list for EDA is to add general support for HyperLinkedFields (without the logic to build the url with the query param). Does this sound reasonable to you?

**Comment by dustinfarris** [comment]
I am going to finish the work on https://github.com/emberjs/data/pull/2468 so we can support coalesceFindRequests.  Hope to do so by the end of the week.

As things are now, (if you remove the block in the adapter) I've noticed that with `API_ADD_TRAILING_SLASHES` set to `false` and with django-filter enabled on DRF, things "just work."  It caught me off guard that this would work so easily.

This doesn't really relate to the hyperlinked fields discussion, but I think hasMany will get a lot easier once that pull request is finished.

**Comment by benkonrath** [comment]
I'm surprised by that as well. How does ED know to build the query param url correctly? Maybe ED is loading all of the records, not just the related records. Did you check the console api requests to see if it's loading the related or all of the data?

**Comment by dustinfarris** [comment]
I was wrong on this.  You are right, ED does build the query param URL correctly, but without a filter on the DRF side, DRF returns everything.  So it works, but is not efficient to say the least.

#68 clears this up by documenting how to enable coalesceFindRequests and add a custom filter to DRF.

**Comment by dustinfarris** [comment]
What do we have to do to finish this?

File: addon/serializers/drf.js (hunk 1 above)
Comment (dustinfarris) [inline addon/serializers/drf.js]: if the relationship.kind is 'hasMany', then we could safely assume that if hash['related_name'] begins with "/" or "http://" that it is a link.

Comment (dustinfarris) [inline addon/serializers/drf.js]: something like,

```js
if (relationship.kind === 'hasMany' && hash[payloadRelKey].match(/^\//)) {
  // Add to links ...
}
```

**Comment by benkonrath** [comment]
I actually have this almost finished in another branch. I'm back to doing Ember stuff again so I'll prioritize getting this done.

**Comment by benkonrath** [comment]
I did a bit of work on this today on another branch (with tests) and it looks like it's not going to need a custom Hyperlinked serializer field. The problem is that the `hasMany` relationship is created with multiple URLs instead of one URL that will limit only return the related items. For example:

Post
```json
{
    "id": 11, 
    "title": "post title 11", 
    "body": "post body 11", 
    "comments": [
        "http://localhost:8000/api/comments/9/", 
        "http://localhost:8000/api/comments/10/", 
        "http://localhost:8000/api/comments/11/"
    ]
}
```

It's possible to support the `belongsTo` relationship in the default DRF configuration because it uses one URL. For example:

Comment
```json
{
    "id": 9, 
    "body": "comment body 1 for post 11", 
    "post": "http://localhost:8000/api/posts/11/"
}
```

The next task would be to create the custom Hyperlinked serializer field that generates one URL using the DjangoFilterBackend or using embedded URLs (e.g. `/api/posts/11/comments/`). Is this something that you think is useful? If not, I'll just drop this and use coalesceFindRequests.

**Comment by benkonrath** [comment]
Closing this PR in favour of #95.

**Pull request closed without merging**